src/channel.rs

```diff
@@ -31,36 +33,49 @@ pub struct Channel {
 #[derive(Debug, Deserialize)]
 pub struct Package {
     pub target: HashMap<String, HashedBinary>,
-    #[serde(deserialize_with = "version_and_date_from_str")]
+    #[serde(deserialize_with = "package_version_from_str")]
     pub version: PackageVersion,
 }
 
-#[derive(Debug)]
+#[derive(Debug, PartialEq, Eq, Ord, PartialOrd)]
 pub struct PackageVersion {
     pub semver: Version,
     pub date: Option<Date>,
 }
 
-fn version_and_date_from_str<'de, D>(deserializer: D) -> Result<PackageVersion, D::Error>
+fn package_version_from_str<'de, D>(deserializer: D) -> Result<PackageVersion, D::Error>
 where
     D: Deserializer<'de>,
 {
     let s = String::deserialize(deserializer)?;
-    let (version, date) = match s.split_once(' ') {
-        Some((version, date)) => (version, Some(date)),
-        None => (s.as_str(), None),
-    };
-
-    let parsed_date = if let Some(date) = date {
-        Date::parse(date, format_description!("([year]-[month]-[day])")).ok()
-    } else {
-        None
-    };
-
-    Ok(PackageVersion {
-        semver: Version::parse(version).unwrap(),
-        date: parsed_date,
-    })
+    s.parse().map_err(serde::de::Error::custom)
+}
+
+impl FromStr for PackageVersion {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut elems = s.split(' ');
+        let semver_str = elems
+            .next()
+            .ok_or_else(|| anyhow!("package version string is empty"))?;
+        let date_str = elems.next();
+        let semver = semver::Version::parse(semver_str).context("failed to parse semver")?;
+        let date = match date_str {
+            None => None,
+            Some(s) => Date::parse(s, format_description!("([year]-[month]-[day])"))
+                .context("specified date is invalid")
```

Comment:
I think we want to return an error here in the case that there is a `date_str` but it can't be parsed, e.g. perhaps
```rust
            Some(s) => {
                let date = Date::parse(s, format_description!("([year]-[month]-[day])"))
                    .context("specified date is invalid")?;
                Some(date)
            }
```
this way we don't silently fall back to a `None` in the case a date was intended but malformed for example.